src/ir/auth_logic/ast_construction.cc

```diff
@@ -78,22 +79,27 @@ static BaseFact ConstructFlatFact(AuthLogicParser::FlatFactContext& context) {
   return BaseFact(ConstructPredicate(*CHECK_NOTNULL(ctx.predicate())));
 }
 
-static Fact ConstructFact(AuthLogicParser::FactContext& context) {
+static Fact ConstructFact(std::forward_list<Principal> delegation_chain,
+                          AuthLogicParser::FactContext& context) {
   if (auto* flat_fact_fact_ctx =
           dynamic_cast<AuthLogicParser::FlatFactFactContext*>(
               &context)) {  // FlatFactFactContext
     return Fact(
+        std::move(delegation_chain),
         ConstructFlatFact(*CHECK_NOTNULL(flat_fact_fact_ctx->flatFact())));
   }
   auto& ctx = *CHECK_NOTNULL(
       dynamic_cast<AuthLogicParser::CanSayFactContext*>(&context));
-  return ConstructFact(*CHECK_NOTNULL(ctx.fact()));
+  delegation_chain.push_front(
+      ConstructPrincipal(*CHECK_NOTNULL(ctx.principal())));
+  return ConstructFact(delegation_chain, *CHECK_NOTNULL(ctx.fact()));
```

Comment:
`delegation_chain` should be `std::move`'d into `ConstructFact` now that `delegation_chain` is a value type. That allows `ConstructFact` to just copy elide `delegation_chain` into `ConstructFact`'s parameter.